pkg/gardenlet/bootstrap/certificate/certificate_rotation.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	bootstraputil "github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// certificateWaitTimeout controls the amount of time we wait for the certificate
+// approval in one iteration.
+var certificateWaitTimeout = 15 * time.Minute
+
+// Manager can be used to schedule the certificate rotation for the Gardenlet's Garden cluster client certificate
+type Manager struct {
+	seedName               string
+	clientMap              clientmap.ClientMap
+	gardenClientConnection *config.GardenClientConnection
+
+	logger logrus.FieldLogger
+	// Set to time.Now but can be stubbed out for testing
+	now func() time.Time
+}
+
+// NewCertificateRotation creates a certificate manager that can be used to rotate gardenlet's client certificate for the Garden cluster
+func NewCertificateManager(gardenClientConnection *config.GardenClientConnection, seedName string, clientMap clientmap.ClientMap) *Manager {
+	return &Manager{
+		seedName:               seedName,
+		clientMap:              clientMap,
+		gardenClientConnection: gardenClientConnection,
+		logger:                 logger.NewFieldLogger(logger.Logger, "certificate-rotation", seedName),
+		now:                    time.Now,
+	}
+}
+
+// ScheduleCertificateRotation waits until the currently used Garden cluster client certificate approaches expiration.
+// Then requests a new certificate and stores the kubeconfig in a secret (`gardenClientConnection.kubeconfigSecret`) on the Seed.
+// the argument is context.Cancel function to cancel the context of the Gardenlet used for graceful termination after a successful certificate rotation.
+// when the new gardenlet pod is started, it uses the rotated certificate stored in the secret in the Seed cluster
+func (cr *Manager) ScheduleCertificateRotation(ctx context.Context, gardenletCancel context.CancelFunc) {
+	go wait.Until(func() {
+		if err := waitForCertificateRotation(ctx, cr.logger, cr.now, cr.gardenClientConnection, cr.clientMap, cr.seedName); err != nil {
+			cr.logger.Errorf("waiting for the certificate rotation failed: %v", err)
+			return
+		}
+
+		err := retry.Until(ctx, certificateWaitTimeout, func(ctx context.Context) (bool, error) {
+			ctxWithTimeout, cancel := context.WithTimeout(ctx, certificateWaitTimeout)
+			defer cancel()
+
+			err := rotateCertificate(ctxWithTimeout, cr.logger, cr.gardenClientConnection, cr.clientMap, cr.seedName)
+			if err != nil {
+				cr.logger.Errorf("certificate rotation failed: %v", err)
+				return retry.MinorError(err)
+			}
+			return retry.Ok()
+		})
+		if err != nil {
+			cr.logger.Errorf("Encountered error when trying to rotate certificate: %v", err)
+			return
```

Comment:
I'm a bit worried that failed renewals remain unnoticed for a long time and at worst lasts until the cert is finally outdated.

Reply:
The gardenlet now emits an event in case the rotation failed.

```
default       6s          Warning   GardenletCertificateRotationFailed   seed/seed-aws-boot                                      Failed to rotate the kubeconfig for the Garden API Server. Certificate expires in 8759h54m48s at (2021-08-11 14:31:07 +0000 UTC): dummy error
```

Though I could not figure out why it is recorded successfully but not shown when `describing` the seed.

Reply:
I guess, this is because you are not passing the `metadata.uid` to the event, which is needed for the association in `k describe seed ...`.
Best option would probably be to have the full Seed object to pass to `recorder.Event`.

Reply:
Yep, working now.